contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,266 @@
+
+    struct UnbondRequest {
+        uint256 sharesAmount;
+        uint256 bnbAmount;
+        uint256 unlockTime;
+    }
+
+    /*----------------- events -----------------*/
+    event Delegated(address indexed sender, uint256 sharesAmount, uint256 bnbAmount);
+    event RewardReceived(uint256 bnbAmount);
+    event UnbondRequested(address indexed sender, uint256 sharesAmount, uint256 bnbAmount, uint256 unlockTime);
+    event UnbondClaimed(address indexed sender, uint256 sharesAmount, uint256 bnbAmount);
+
+    /*----------------- modifiers -----------------*/
+
+    /*----------------- external functions -----------------*/
+    function initialize(address _validator, uint256 minSelfDelegationBNB) public payable initializer {
+        __EIP712_init_unchained(EIP712_NAME, EIP712_VERSION);
+        __ERC20_init_unchained(ERC20_NAME, ERC20_SYMBOL);
+
+        validator = _validator;
+
+        assert(msg.value != 0);
+        _bootstrapInitialHolder(msg.value);
+
+        // transfer to stakeHub
+        _transfer(validator, STAKE_HUB_ADDR, minSelfDelegationBNB);
+    }
+
+    function delegate(address _delegator) external payable onlyStakeHub returns (uint256) {
+        require(msg.value != 0, "ZERO_DEPOSIT");
+        return _stake(_delegator, msg.value);
+    }
+
+    function undelegate(address _delegator, uint256 _sharesAmount) external onlyStakeHub returns (uint256) {
+        require(_sharesAmount != 0, "ZERO_AMOUNT");
+        require(_sharesAmount <= balanceOf(_delegator), "INSUFFICIENT_BALANCE");
+
+        _lockedShares[_delegator] += _sharesAmount;
+
+        // calculate the BNB amount and update state
+        uint256 _bnbAmount = getPooledBNBByShares(_sharesAmount);
+        _burn(_delegator, _sharesAmount);
+        _totalPooledBNB -= _bnbAmount;
+
+        // add to the queue
+        _unbondSequence[_delegator] += 1; // increase the sequence first to avoid zero sequence
+        bytes32 hash = keccak256(abi.encodePacked(_delegator, _unbondSequence[_delegator]));
+
+        uint256 unlockTime = block.timestamp + IStakeHub(STAKE_HUB_ADDR).unbondPeriod();
+        UnbondRequest memory request = UnbondRequest({sharesAmount: _sharesAmount, bnbAmount: _bnbAmount, unlockTime: unlockTime});
+        _unbondRequests[hash] = request;
+        _unbondRequestsQueue[_delegator].pushBack(hash);
+
+        emit UnbondRequested(_delegator, _sharesAmount, _bnbAmount, request.unlockTime);
+        return _bnbAmount;
+    }
+
+    function redelegate(address _delegator, uint256 _sharesAmount) external onlyStakeHub returns (uint256) {
+        require(_sharesAmount <= balanceOf(_delegator), "INSUFFICIENT_BALANCE");
+
+        // calculate the BNB amount and update state
+        uint256 _bnbAmount = getPooledBNBByShares(_sharesAmount);
+        _burn(_delegator, _sharesAmount);
+        _totalPooledBNB -= _bnbAmount;
+
+        (bool success,) = STAKE_HUB_ADDR.call{value: _bnbAmount}("");
```

Comment:
Limit the gas here as well, even though it is our trustworthy contract.

Reply:
Done